# Разработка дашборда для благотворительного фонда "Рауль", Санкт-Петербург

## 1. Введение.

**Благотворительный фонд «Рауль»** поддерживает молодых людей с инвалидностью, особенностями здоровья и сиротским опытом, помогает им научиться самостоятельной жизни, осознанно выбрать себе учебу и работу и содействует развитию условий и инфраструктуры для этого в регионах.

Миссия фонда «Рауль» — помогать молодым людям с низкими стартовыми возможностями достичь большей независимости, реализовать свое право на работу, развитие и полноценную жизнь.

**Программы фонда.**

**ПРОГРАММА ВСЁ ПОЛУЧИТСЯ!**\
vsepoluchitsya.org

Программа «Всё получится!» направлена на создание условий и инфраструктуры для эффективной социальной адаптации и трудоустройства молодых людей с низкими стартовыми возможностями.
Для этого программа объединяет усилия партнеров - бизнеса, некоммерческих организаций и региональных служб занятости.

В рамках программы созданы материалы для некоммерческих организаций и бизнеса:
- Описание технологии сопровождаемого трудоустройства;
- Социальная франшиза (для некоммерческих организаций);
- Материалы об инклюзивном найме (для работодателей)

**ПРОГРАММА РАБОТА-i**\
rabota-i.org

Для соискателей — это возможность трудоустройства в российские и международные компании, тренинги, производственные экскурсии, кратковременный стажировки и практические занятия по поиску работы. Ключевая практика — сопровождаемое трудоустройство.

Для работодателей — разнообразные КСО-программы в партнерстве с фондами и общественными организациями, вовлечение корпоративных волонтёров в поддержку молодых людей с низкими стартовыми возможностями.

**Задание для аналитиков** - построить дэшборд. 
Дэшборд располагается на ресурсе, доступном сотрудникам организации БФ Рауль (https://raoul.spb.ru/).
Дэшборд отображает динамику и структуру трудоустройств.

**Параметры динамики:** помесячно, поквартально, ежегодно.

**Параметры структуры:** принадлежность благополучателя к одному из сегментов, ответственная НКО, регион, пол благополучателя, возраст, тип трудоустройства (временное, постоянное с сопровождением, постоянное без сопровождения), тип работодателя (внутренняя защищённая площадка, внешняя защищённая площадка, открытый рынок).

**Параметры трудоустройства:** от 1 дня, от 1 месяца
Back-end дэшборда можно редактировать без привлечения внешних специалистов (например, этим может заниматься специалист по мониторингу и оценке)


## 2. Содержание проекта.

1.Загрузка данных и их подготовка:
- загрузка данных и ознакомление с датасетами;
- обработка пропусков;
- проверка соответствия типов данных и их изменение;
- обработка дубликатов;
- посмотреть на наличие ошибок в данных.
2.Исследование данных.\
3.Построение дашборда.\
4.Выводы и рекомендации.

## 3. Загрузка данных и ознакомление с датасетами.

In [1]:
#загружаем необходимые библиотеки
import pandas as pd
import datetime

Загружаем данные:

In [2]:
raul_df = pd.read_excel('C:\\Users\\77281\\Downloads\\БФРауль_Данные_по_трудоустроенным.xls')

Первичное ознакомление с датасетом:

In [3]:
raul_df.head()

Unnamed: 0  id_кандидата      пол  возраст дата_добавления инвалидность  \
0           0             1  Женский       25      2018-12-06       ИСТИНА   
1           1             2  Мужской       28      2021-09-22       ИСТИНА   
2           2             3  Женский       31      2016-07-08       ИСТИНА   
3           3             4  Мужской       34      2022-10-07       ИСТИНА   
4           4             4  Мужской       34      2022-10-07       ИСТИНА   

  сиротский_опыт адаптированная_образовательная_программа  \
0           ЛОЖЬ                                     ЛОЖЬ   
1           ЛОЖЬ                                     ЛОЖЬ   
2           ЛОЖЬ                                     ЛОЖЬ   
3           ЛОЖЬ                                     ЛОЖЬ   
4           ЛОЖЬ                                     ЛОЖЬ   

  начало_трудоустройства завершение_трудоустройства 1+_месяц_работы  \
0             2023-01-16                        NaT          ИСТИНА   
1             2023-01-09                        NaT          ИСТИНА   
2             2023-04-18                 2023-06-30          ИСТИНА   
3             2022-11-21                 2022-12-30          ИСТИНА   
4             2023-03-22                 2023-03-30            ЛОЖЬ   

  тип_площадки               тип_трудоустройства тип_опыта_работы регион_код  \
0      Внешняя  Открытый рынок без сопровождения        Повторный     reg200   
1      Внешняя   Открытый рынок с сопровождением        Повторный     reg200   
2      Внешняя  Открытый рынок без сопровождения        Повторный     reg200   
3      Внешняя  Открытый рынок без сопровождения        Повторный     reg200   
4      Внешняя  Открытый рынок без сопровождения        Повторный     reg200   

  провайдер_код  
0      prov3500  
1      prov3500  
2      prov3500  
3      prov3500  
4      prov3500

In [4]:
raul_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 259 entries, 0 to 258
Data columns (total 16 columns):
 #   Column                                    Non-Null Count  Dtype         
---  ------                                    --------------  -----         
 0   Unnamed: 0                                259 non-null    int64         
 1   id_кандидата                              259 non-null    int64         
 2   пол                                       259 non-null    object        
 3   возраст                                   259 non-null    int64         
 4   дата_добавления                           258 non-null    datetime64[ns]
 5   инвалидность                              259 non-null    object        
 6   сиротский_опыт                            259 non-null    object        
 7   адаптированная_образовательная_программа  259 non-null    object        
 8   начало_трудоустройства                    259 non-null    datetime64[ns]
 9   завершение_трудоустройства      

В датасете имеется информации о 259 кандидатах; все данные в столбцах находятся в соответственных типах данных. В нескольких столбцах видим пропуски.

## 4. Предобработка данных.

Считаем пропуски:

In [5]:
raul_df.isna().sum()

Unnamed: 0                                    0
id_кандидата                                  0
пол                                           0
возраст                                       0
дата_добавления                               1
инвалидность                                  0
сиротский_опыт                                0
адаптированная_образовательная_программа      0
начало_трудоустройства                        0
завершение_трудоустройства                  174
1+_месяц_работы                               0
тип_площадки                                  1
тип_трудоустройства                           1
тип_опыта_работы                              1
регион_код                                    0
провайдер_код                                 0
dtype: int64

Основные пропуски в столбце 'завершение_трудоустройства', они говорят о том, что кандидат продолжает работу на последнем месте трудоустройства, поэтому вместо пропущенного значения вставляем сегодняшнюю дату:

In [6]:
curr_date = pd.Timestamp('today').strftime("%Y-%m-%d")
curr_date

'2023-11-25'

In [7]:
raul_df['завершение_трудоустройства'].fillna(curr_date, inplace = True)
raul_df.head()

Unnamed: 0  id_кандидата      пол  возраст дата_добавления инвалидность  \
0           0             1  Женский       25      2018-12-06       ИСТИНА   
1           1             2  Мужской       28      2021-09-22       ИСТИНА   
2           2             3  Женский       31      2016-07-08       ИСТИНА   
3           3             4  Мужской       34      2022-10-07       ИСТИНА   
4           4             4  Мужской       34      2022-10-07       ИСТИНА   

  сиротский_опыт адаптированная_образовательная_программа  \
0           ЛОЖЬ                                     ЛОЖЬ   
1           ЛОЖЬ                                     ЛОЖЬ   
2           ЛОЖЬ                                     ЛОЖЬ   
3           ЛОЖЬ                                     ЛОЖЬ   
4           ЛОЖЬ                                     ЛОЖЬ   

  начало_трудоустройства завершение_трудоустройства 1+_месяц_работы  \
0             2023-01-16                 2023-11-25          ИСТИНА   
1             2023-01-09                 2023-11-25          ИСТИНА   
2             2023-04-18                 2023-06-30          ИСТИНА   
3             2022-11-21                 2022-12-30          ИСТИНА   
4             2023-03-22                 2023-03-30            ЛОЖЬ   

  тип_площадки               тип_трудоустройства тип_опыта_работы регион_код  \
0      Внешняя  Открытый рынок без сопровождения        Повторный     reg200   
1      Внешняя   Открытый рынок с сопровождением        Повторный     reg200   
2      Внешняя  Открытый рынок без сопровождения        Повторный     reg200   
3      Внешняя  Открытый рынок без сопровождения        Повторный     reg200   
4      Внешняя  Открытый рынок без сопровождения        Повторный     reg200   

  провайдер_код  
0      prov3500  
1      prov3500  
2      prov3500  
3      prov3500  
4      prov3500

Единичный пропуск в столбце'дата_добавления' заменим на значение столбца 'начало_трудоустройства'.

In [8]:
raul_df['дата_добавления'].fillna(raul_df['начало_трудоустройства'],inplace = True)

Так же видим по одному пропуску в столбцах 'тип_площадки', 'тип_трудоустройства' и 'тип_опыта_работы'.
Как пояснил заказчик, это не должно мешать считать этих трудоустроенных, потому что первостепенно важные данные - кандидат, провайдер, регион, начало трудоустройства для этих кандидатов имеются. Поэтому оставим их без изменения.

Все пропуски в датафрейме обработаны.

Считаем явные дубликаты:

In [9]:
raul_df.duplicated().sum()

0

Явных дубликатов не обнаружено.

In [10]:
raul_df.to_csv('raul_df.csv')

Проверим данные на наличие ошибок. Участие в программе фонда подразумевает наличие хотя бы одного из условий: инвалидность, сиротский опыт или прохождение адаптивной образовательной программы.

In [11]:
print(raul_df.query('инвалидность=="ЛОЖЬ" and сиротский_опыт=="ЛОЖЬ" and адаптированная_образовательная_программа=="ЛОЖЬ"'))

     Unnamed: 0  id_кандидата      пол  возраст дата_добавления инвалидность  \
66           66            62  Мужской       21      2018-05-16         ЛОЖЬ   
210         210           198  Женский       16      2023-08-05         ЛОЖЬ   
211         211           199  Мужской       14      2023-06-13         ЛОЖЬ   
213         213           201  Мужской       16      2023-06-13         ЛОЖЬ   
219         219           207  Женский       16      2023-06-27         ЛОЖЬ   
220         220           208  Женский       14      2023-01-10         ЛОЖЬ   

    сиротский_опыт адаптированная_образовательная_программа  \
66            ЛОЖЬ                                     ЛОЖЬ   
210           ЛОЖЬ                                     ЛОЖЬ   
211           ЛОЖЬ                                     ЛОЖЬ   
213           ЛОЖЬ                                     ЛОЖЬ   
219           ЛОЖЬ                                     ЛОЖЬ   
220           ЛОЖЬ                                     ЛОЖЬ  

Про кандидата с id62 на встрече с заказчиком выявили ошибку в данных - у кандидата инвалидность и адаптированная образовательная программа. Внесем изменения в данные, по кандидатам с id 210,211,213,219,220 вопрос к заказчику о проверке данных остается в силе.

In [12]:
raul_df.loc[raul_df['id_кандидата'] == 62, 'инвалидность'] = 'ИСТИНА'
raul_df.loc[raul_df['id_кандидата'] == 62, 'адаптированная_образовательная_программа'] = 'ИСТИНА'

In [13]:
raul_df.loc[raul_df['id_кандидата'] == 62]

Unnamed: 0  id_кандидата      пол  возраст дата_добавления инвалидность  \
65          65            62  Мужской       21      2018-05-16       ИСТИНА   
66          66            62  Мужской       21      2018-05-16       ИСТИНА   

   сиротский_опыт адаптированная_образовательная_программа  \
65           ЛОЖЬ                                   ИСТИНА   
66           ЛОЖЬ                                   ИСТИНА   

   начало_трудоустройства завершение_трудоустройства 1+_месяц_работы  \
65             2023-04-24                 2023-06-30          ИСТИНА   
66             2023-08-24                 2023-11-25            ЛОЖЬ   

   тип_площадки               тип_трудоустройства тип_опыта_работы регион_код  \
65      Внешняя         Временное трудоустройство        Повторный     reg200   
66      Внешняя  Открытый рынок без сопровождения        Повторный     reg200   

   провайдер_код  
65      prov3500  
66      prov3500

In [14]:
raul_df.head()

Unnamed: 0  id_кандидата      пол  возраст дата_добавления инвалидность  \
0           0             1  Женский       25      2018-12-06       ИСТИНА   
1           1             2  Мужской       28      2021-09-22       ИСТИНА   
2           2             3  Женский       31      2016-07-08       ИСТИНА   
3           3             4  Мужской       34      2022-10-07       ИСТИНА   
4           4             4  Мужской       34      2022-10-07       ИСТИНА   

  сиротский_опыт адаптированная_образовательная_программа  \
0           ЛОЖЬ                                     ЛОЖЬ   
1           ЛОЖЬ                                     ЛОЖЬ   
2           ЛОЖЬ                                     ЛОЖЬ   
3           ЛОЖЬ                                     ЛОЖЬ   
4           ЛОЖЬ                                     ЛОЖЬ   

  начало_трудоустройства завершение_трудоустройства 1+_месяц_работы  \
0             2023-01-16                 2023-11-25          ИСТИНА   
1             2023-01-09                 2023-11-25          ИСТИНА   
2             2023-04-18                 2023-06-30          ИСТИНА   
3             2022-11-21                 2022-12-30          ИСТИНА   
4             2023-03-22                 2023-03-30            ЛОЖЬ   

  тип_площадки               тип_трудоустройства тип_опыта_работы регион_код  \
0      Внешняя  Открытый рынок без сопровождения        Повторный     reg200   
1      Внешняя   Открытый рынок с сопровождением        Повторный     reg200   
2      Внешняя  Открытый рынок без сопровождения        Повторный     reg200   
3      Внешняя  Открытый рынок без сопровождения        Повторный     reg200   
4      Внешняя  Открытый рынок без сопровождения        Повторный     reg200   

  провайдер_код  
0      prov3500  
1      prov3500  
2      prov3500  
3      prov3500  
4      prov3500

Вывод: данные проверены и готовы к дальнейшей работе.

Добавим столбец с длительностью трудоустройства для каждого кандидата:

In [15]:
raul_df['длительность_трудоустройства'] = raul_df['завершение_трудоустройства'] - raul_df['начало_трудоустройства']

Получили столбец в формате данных Timedelta. Приведем формат к integer:

In [16]:
raul_df['длительность_трудоустройства'] = pd.to_numeric(raul_df['длительность_трудоустройства'].dt.days, downcast='integer')

Напишем функцию для распределения кандидатов по группам длительности трудоустройства:

In [17]:
def длительность (row):
        if row['длительность_трудоустройства'] < 1:
            return 'меньше одного дня'
        if row['длительность_трудоустройства'] >= 1 and row['длительность_трудоустройства']<=30:
            return 'от одного дня до месяца'
        if row['длительность_трудоустройства'] > 30 and row['длительность_трудоустройства']<=182:
            return 'от одного месяца до полугода'
        return 'больше полугода'
               
raul_df['длительность'] = raul_df.apply(длительность, axis = 1)

In [19]:
def от_одного_дня (row):
        if row['длительность_трудоустройства'] > 1:
            return 'истина'
        return 'ложь'
               
raul_df['от_одного_дня'] = raul_df.apply(от_одного_дня, axis = 1)

In [22]:
def от_одного_месяца (row):
        if row['длительность_трудоустройства'] > 30:
            return 'истина'
        return 'ложь'
               
raul_df['от_одного_месяца'] = raul_df.apply(от_одного_месяца, axis = 1)

In [23]:
def от_трех_месяцев (row):
        if row['длительность_трудоустройства'] > 91:
            return 'истина'
        return 'ложь'
               
raul_df['от_трех_месяцев'] = raul_df.apply(от_трех_месяцев, axis = 1)

In [24]:
def от_полугода (row):
        if row['длительность_трудоустройства'] > 182:
            return 'истина'
        return 'ложь'
               
raul_df['от_полугода'] = raul_df.apply(от_полугода, axis = 1)

In [25]:
raul_df

Unnamed: 0  id_кандидата      пол  возраст дата_добавления инвалидность  \
0             0             1  Женский       25      2018-12-06       ИСТИНА   
1             1             2  Мужской       28      2021-09-22       ИСТИНА   
2             2             3  Женский       31      2016-07-08       ИСТИНА   
3             3             4  Мужской       34      2022-10-07       ИСТИНА   
4             4             4  Мужской       34      2022-10-07       ИСТИНА   
..          ...           ...      ...      ...             ...          ...   
254         254           240  Мужской       23      2022-08-29       ИСТИНА   
255         255           241  Мужской       22      2022-08-26       ИСТИНА   
256         256           241  Мужской       22      2022-08-26       ИСТИНА   
257         257           242  Женский       20      2022-08-27       ИСТИНА   
258         258           242  Женский       20      2022-08-27       ИСТИНА   

    сиротский_опыт адаптированная_образовательная_программа  \
0             ЛОЖЬ                                     ЛОЖЬ   
1             ЛОЖЬ                                     ЛОЖЬ   
2             ЛОЖЬ                                     ЛОЖЬ   
3             ЛОЖЬ                                     ЛОЖЬ   
4             ЛОЖЬ                                     ЛОЖЬ   
..             ...                                      ...   
254           ЛОЖЬ                                     ЛОЖЬ   
255           ЛОЖЬ                                     ЛОЖЬ   
256           ЛОЖЬ                                     ЛОЖЬ   
257           ЛОЖЬ                                     ЛОЖЬ   
258           ЛОЖЬ                                     ЛОЖЬ   

    начало_трудоустройства завершение_трудоустройства  ...  \
0               2023-01-16                 2023-11-25  ...   
1               2023-01-09                 2023-11-25  ...   
2               2023-04-18                 2023-06-30  ...   
3               2022-11-21                 2022-12-30  ...   
4               2023-03-22                 2023-03-30  ...   
..                     ...                        ...  ...   
254             2023-04-18                 2023-11-25  ...   
255             2023-02-27                 2023-03-10  ...   
256             2023-03-21                 2023-11-25  ...   
257             2022-12-12                 2022-12-30  ...   
258             2023-06-22                 2023-11-25  ...   

                  тип_трудоустройства тип_опыта_работы регион_код  \
0    Открытый рынок без сопровождения        Повторный     reg200   
1     Открытый рынок с сопровождением        Повторный     reg200   
2    Открытый рынок без сопровождения        Повторный     reg200   
3    Открытый рынок без сопровождения        Повторный     reg200   
4    Открытый рынок без сопровождения        Повторный     reg200   
..                                ...              ...        ...   
254   Открытый рынок с сопровождением           Первый     reg209   
255         Временное трудоустройство           Первый     reg209   
256   Открытый рынок с сопровождением        Повторный     reg209   
257         Временное трудоустройство           Первый     reg209   
258  Открытый рынок без сопровождения        Повторный     reg209   

    провайдер_код длительность_трудоустройства                  длительность  \
0        prov3500                          313               больше полугода   
1        prov3500                          320               больше полугода   
2        prov3500                           73  от одного месяца до полугода   
3        prov3500                           39  от одного месяца до полугода   
4        prov3500                            8       от одного дня до месяца   
..            ...                          ...                           ...   
254      prov3507                          221               больше полугода   
255      prov3507                           11       от одного дня до мес

Сохраняем файл для экспорта в DataLens для построения дашборда:

In [26]:
raul_df.to_csv('C:\\Users\\77281\\raul_df.csv', index= False)